DoublyLinkedList/implementation.cpp

```diff
@@ -0,0 +1,219 @@
+            head = node;
+        }
+        this->sz++;
+    }
+
+    void addLast(int val){
+        Node* node= new Node(val);
+        if(this->head == nullptr){
+            this->head = node; this->tail = node;
+        }else if(head == tail){
+            head->next = node; node->prev = head;
+            tail = node;
+        }else{
+            tail->next = node; node->prev = tail;
+            tail = node;
+        }
+        this->sz++;
+    }
+    void addAt(int val,int idx){
+        if(idx == 0){
+            addFirst(val);
+        }else if(idx == this->sz){
+            addLast(val);
+        }else{
+            int i = 0;
+            Node* temp = head;
+            while(i<idx){
+                temp = temp->next;
+                i++;
+            }
+            Node* newNode = new Node(val);
+            temp->prev->next = newNode;
+            newNode->prev = temp->prev;
+            newNode->next = temp;
+            temp->prev = newNode;
+        }
+        this->sz++;
+    }
+
+    // REMOVE
+    int removeFirst(){
+        if(head == nullptr){
+            cout<<"List is empty"<<endl;
+            return -1;
+        }else if(head == tail){
+            int ret = head->data;
+            head = tail = nullptr; this->sz--;
+            return ret;
+        }else{
+            int ret = head->data;
+            head = head->next;
+            head->prev = nullptr;
+            this->sz--; return ret;
+        }
+    }
+
+    int removeLast(){
+        if(head == nullptr){
+            cout<<"List is empty"<<endl;
+            return -1;
+        }else if(head == tail){
+            int ret = head->data;
+            head = tail = nullptr; this->sz--;
+            return ret;
+        }else{
+            int ret = tail->data;
+            tail = tail->prev; tail->next = nullptr;
+            this->sz--; return ret;
+        }
+    }
+    
+    int removeAt(int idx){
+        if(idx<0 || idx>=this->sz){
+            cout<<"Enter valid index"<<endl;
+            return -1;
+        }
+        if(idx == 0){
+            return removeFirst();
+        }else if(idx == this->sz-1){
+            return removeLast();
+        }else{
+            int i = 0;
+            Node* temp = head;
+            while(i<idx){
+                temp = temp->next;
+                i++;
+            }
+            int ret = temp->data;
+            temp->prev->next = temp->next;
+            temp->next->prev = temp->prev;
+            temp = nullptr;
+            this->sz--;
+            return ret;
+        }
+    }
+
+    Node* getAt(int idx){
+        if(idx == 0){
+            return this->head;
+        }else if(idx == this->sz-1){
+            return this->tail;
+        }else{
+            int i = 0;
+            Node* temp = head;
+            while(i<idx){
+                temp = temp->next;i++;
+            }
+            return temp;
+        }
+    }
+};
+
+
+
+int32_t main(){
+//     ios::sync_with_stdio(0); cin.tie(0);
+    
```

Comment:
why is the main() empty? shouldn't it be creating a DoublyLinkedList object in order to execute the program? pls correct it @Rohanfizz

Reply:
It would work just fine, I've written it this way because I've #defined long long int to int to prevent integer overflows, If you still want it normal I will change immediately :)